pkg/sdn/plugin/sdn-cni-plugin/openshift-sdn.go

```diff
@@ -0,0 +1,511 @@
+		parts := strings.Split(arg, "=")
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("invalid CNI_ARG '%s'", arg)
+		}
+		mapArgs[strings.TrimSpace(parts[0])] = strings.TrimSpace(parts[1])
+	}
+	return mapArgs, nil
+}
+
+func readPodInfo(originClient *oclient.Client, kubeClient *kclient.Client, cniArgs string, pluginName string) (*podInfo, error) {
+	mapArgs, err := gatherCniArgs(cniArgs)
+	if err != nil {
+		return nil, err
+	}
+
+	namespace, namespaceOk := mapArgs["K8S_POD_NAMESPACE"]
+	name, nameOk := mapArgs["K8S_POD_NAME"]
+	if !namespaceOk || !nameOk {
+		return nil, fmt.Errorf("missing pod namespace or name")
+	}
+
+	info := &podInfo{}
+
+	if sdnplugin.IsOpenShiftMultitenantNetworkPlugin(pluginName) {
+		netNamespace, err := originClient.NetNamespaces().Get(namespace)
+		if err != nil {
+			return nil, err
+		}
+		info.Vnid = netNamespace.NetID
+	}
+
+	// FIXME: does this ensure the returned pod lives on this node?
+	pod, err := kubeClient.Pods(namespace).Get(name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read pod %s/%s: %v", namespace, name, err)
+	}
+
+	for _, container := range pod.Spec.Containers {
+		if container.SecurityContext.Privileged != nil && *container.SecurityContext.Privileged {
+			info.Privileged = true
+			break
+		}
+	}
+
+	info.Annotations = pod.Annotations
+	return info, nil
+}
+
+func getBandwidth(pi *podInfo) (string, string, error) {
+	ingress, egress, err := kbandwidth.ExtractPodBandwidthResources(pi.Annotations)
+	if err != nil {
+		return "", "", fmt.Errorf("failed to parse pod bandwidth: %v", err)
+	}
+	var ingressStr, egressStr string
+	if ingress != nil {
+		ingressStr = fmt.Sprintf("%d", ingress.Value())
+	}
+	if egress != nil {
+		egressStr = fmt.Sprintf("%d", egress.Value())
+	}
+	return ingressStr, egressStr, nil
+}
+
+func wantsMacvlan(pi *podInfo) (bool, error) {
+	val, found := pi.Annotations[AssignMacVlanAnnotation]
+	if !found || val != "true" {
+		return false, nil
+	}
+	if pi.Privileged {
+		return true, nil
+	}
+	return false, fmt.Errorf("pod has %q annotation but is not privileged", AssignMacVlanAnnotation)
+}
+
+func loadNetConf(bytes []byte) (*api.CNINetConfig, error) {
+	n := &api.CNINetConfig{}
+	if err := json.Unmarshal(bytes, n); err != nil {
+		return nil, fmt.Errorf("failed to load netconf: %v", err)
+	}
+	return n, nil
+}
+
+func getPodConfig(args *skel.CmdArgs) (*PodConfig, error) {
+	n, err := loadNetConf(args.StdinData)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create OpenShift and Kubernetes clients to read the ClusterNetwork and PodSpec
+	overrides := &configapi.ClientConnectionOverrides{}
+	configapi.SetProtobufClientDefaults(overrides)
+	originClient, _, err := configapi.GetOpenShiftClient(n.MasterKubeConfig, overrides)
+	if err != nil {
+		return nil, err
+	}
+	kubeClient, _, err := configapi.GetKubeClient(n.MasterKubeConfig, overrides)
+	if err != nil {
+		return nil, err
+	}
+
+	// Grab the node's network configuration
```

Comment:
Yes, Clayton wanted it read from master per https://github.com/dcbw/openshift-sdn/commit/62088467cc94420cd624fdd1f91a43bc29055074#commitcomment-18067709 and the next two.

There is another possibility: that our CNI plugin does little-to-nothing and simply passes the CNI args back to the openshift-sdn process running on the node, which does the container setup.  That makes HostPort easier actually.  But it would involve private API that Clayton clearly disliked in https://github.com/openshift/openshift-sdn/pull/320#discussion_r68142557.